crucible-mir/src/Mir/Intrinsics.hs

```diff
@@ -801,6 +804,121 @@ readMirAggregateWithSymOffset bak iteFn off tpr ag@(MirAggregate totalSize m)
     offsetLit = wordLit sym
     iteFn' = liftIteFnMaybe sym tpr iteFn
 
+    -- Given a list of valid offsets in this aggregate, create a predicate that
+    -- the provided offset appears among them.
+    --
+    -- Precondition: the candidate offsets are sorted in ascending order.
+    offsetIn ::
+      NonEmpty Word ->
+      MuxLeafT sym IO (SymExpr sym BaseBoolType)
+    offsetIn (off0 :| offs) = liftIO $ do
+      -- We want to construct a symbolic expression representing whether or not
+      -- `off` appears in this aggregate's candidate offsets. The most
+      -- straightforward way to do that is to test it for equality against each
+      -- of those offsets and `orPred` those tests into a single expression.
+      -- However, expressions like that can be expensive to simulate.
+      --
+      -- If portions of the aggregate represent arrays, then those entries will
+      -- be evenly spaced a known stride apart from one another. If we can find
+      -- such contiguous "runs" of entries, we can construct much simpler
+      -- membership predicates for them: is the provided offset between the
+      -- beginning and the end of a run, and is the offset a multiple of the
+      -- stride of that run?
+      --
+      -- We partition the aggregate into contiguous runs and discrete elements
+      -- to take advantage of the more efficient predicate where possible, and
+      -- to fall back to the less efficient predicate where necessary.
+      let spans = List.zipWith (,) (off0 : offs) (offs <> [totalSize])
```

Comment:
Another point related to flattened `MirAggregate`s: I don't think it's right to consult `totalSize` here.  Consider:

```Rust
#[repr(C)]
struct S2 {
  x: [u16; 3],
  y: u8,
}
```

When the access type is `BVRepr 16`, the candidate offsets will be 0,2,4, but the `totalSize` will be 8 (not 6).  And we want the three elements of `x` to be a single range in cases like this.

Reply:
You're right - 0891d5717a7cf4d216263845f34ef8bbe764f83a, which addressed your comment about stricter run construction rules, also fixes this by getting the end of the last element directly from the aggregate entry, rather than assuming it's the same as the end of the aggregate.